### ISSUE TITLE
Conflicts between cobrapy and optlang w/ symengine 

### ISSUE THREAD

**Issue by pstjohn** [opencobra/cobrapy#606]
So optlang will pull symengine by default if it's installed, which causes some problems when `cobra` uses sympy objects.

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-18-fa058ddf7116> in <module>()
      7         model.reactions.get_by_id(rid).bounds = (flux, flux)
      8 
----> 9     sol = cobra.flux_analysis.pfba(model)

~/Packages/cobrapy/cobra/flux_analysis/parsimonious.py in pfba(model, already_irreversible, fraction_of_optimum, solver, desired_objective_value, objective, reactions, **optimize_kwargs)
     97         return _pfba_optlang(
     98             model, objective=objective,
---> 99             fraction_of_optimum=fraction_of_optimum, reactions=reactions)
    100 
    101 

~/Packages/cobrapy/cobra/flux_analysis/parsimonious.py in _pfba_optlang(model, objective, reactions, fraction_of_optimum)
    162     with model as m:
    163         add_pfba(m, objective=objective,
--> 164                  fraction_of_optimum=fraction_of_optimum)
    165         m.slim_optimize(error_value=None)
    166         solution = get_solution(m, reactions=reactions)

~/Packages/cobrapy/cobra/flux_analysis/parsimonious.py in add_pfba(model, objective, fraction_of_optimum)
    131     pfba_objective = model.problem.Objective(add(
    132         [mul((sympy.singleton.S.One, variable))
--> 133          for variable in variables]), direction='min', sloppy=True,
    134         name="_pfba_objective")
    135     set_objective(model, pfba_objective)

~/Packages/cobrapy/cobra/flux_analysis/parsimonious.py in <listcomp>(.0)
    131     pfba_objective = model.problem.Objective(add(
    132         [mul((sympy.singleton.S.One, variable))
--> 133          for variable in variables]), direction='min', sloppy=True,
    134         name="_pfba_objective")
    135     set_objective(model, pfba_objective)

~/anaconda3/envs/pymc/lib/python3.6/site-packages/sympy/core/operations.py in _from_args(cls, args, is_commutative)
     58         obj = super(AssocOp, cls).__new__(cls, *args)
     59         if is_commutative is None:
---> 60             is_commutative = fuzzy_and(a.is_commutative for a in args)
     61         obj.is_commutative = is_commutative
     62         return obj

~/anaconda3/envs/pymc/lib/python3.6/site-packages/sympy/core/logic.py in fuzzy_and(args)
    137 
    138     rv = True
--> 139     for ai in args:
    140         ai = fuzzy_bool(ai)
    141         if ai is False:

~/anaconda3/envs/pymc/lib/python3.6/site-packages/sympy/core/operations.py in <genexpr>(.0)
     58         obj = super(AssocOp, cls).__new__(cls, *args)
     59         if is_commutative is None:
---> 60             is_commutative = fuzzy_and(a.is_commutative for a in args)
     61         obj.is_commutative = is_commutative
     62         return obj

AttributeError: 'Variable' object has no attribute 'is_commutative'
```

We'll probably want to make sure we pull all those files with:
```python
from optlang.symbolics import add, mul, One, Zero
```
to make sure we keep compatibility.

Stuff we'll probably have to fix if we want to keep symengine compatibility:
```
$ git grep '\(from\|import\) sympy'
cobra/core/model.py:import sympy
cobra/core/model.py:from sympy import S
cobra/flux_analysis/gapfilling.py:from sympy import Add
cobra/flux_analysis/loopless.py:from sympy.core.singleton import S
cobra/flux_analysis/moma.py:from sympy.core.singleton import S
cobra/flux_analysis/parsimonious.py:import sympy
cobra/flux_analysis/sampling.py:from sympy.core.singleton import S
cobra/flux_analysis/variability.py:from sympy.core.singleton import S
cobra/io/sbml3.py:    from sympy import Basic
cobra/solvers/cglpk.pyx:    from sympy import Basic, Number
cobra/solvers/cplex_solver.py:    from sympy import Basic, Number
cobra/solvers/gurobi_solver.py:    from sympy import Basic, Number
cobra/test/test_model.py:from sympy import S
cobra/util/solver.py:import sympy

$ git grep 'sympy\.'
cobra/core/model.py:        if isinstance(value, sympy.Basic):
cobra/flux_analysis/loopless.py:from sympy.core.singleton import S
cobra/flux_analysis/moma.py:from sympy.core.singleton import S
cobra/flux_analysis/parsimonious.py:add = sympy.Add._from_args
cobra/flux_analysis/parsimonious.py:mul = sympy.Mul._from_args
cobra/flux_analysis/parsimonious.py:        [mul((sympy.singleton.S.One, variable))
cobra/flux_analysis/sampling.py:from sympy.core.singleton import S
cobra/flux_analysis/variability.py:from sympy.core.singleton import S
cobra/util/solver.py:    expression : sympy.Basic
cobra/util/solver.py:            e.g. optlang.glpk_interface.Objective, sympy.Basic or dict
cobra/util/solver.py:                sympy.S.Zero, direction=model.solver.objective.direction)
cobra/util/solver.py:    elif isinstance(value, (sympy.Basic, optlang.interface.Objective)):
cobra/util/solver.py:        if isinstance(value, sympy.Basic):
documentation_builder/conf.py:                'optlang', 'optlang.interface', 'sympy', 'sympy.core',
documentation_builder/conf.py:                'sympy.core.singleton', 'future', 'future.utils', 'ruamel',
```

**Comment by Midnighter** [comment]
Thank you for this investigation!

**Comment by cdiener** [comment]
Does symengine support all the sympy primitives we import? In this case it would be a pretty straight-forward fix (changing all the imports to optlang.symbolics)...

**Comment by pstjohn** [comment]
They've got `One`, `Zero`, `NegativeOne`, `Add`, `Mul`, and `Pow` at least. Do we use anything other than that? @KristianJensen would probably know the best whether symengine supports it.

**Comment by KristianJensen** [comment]
Symengine will crash if sympy objects are mixed in, such as sympy.One etc. As Christian said, the fix is to replace such uses with the objects found in `optlang.symbolics`, e.g. `from optlang.symbolics import One`. This will dynamically import the objects from symengine or sympy so they match the rest of the objects. If there are any of these objects that are not implemented in optlang.symbolics, let me know and I can add them. Note that in addition to the sympy/symengine `Add` and `Mul` objects, `optlang.symbolics` also implements `add` and `mul` functions that will add or multiply the arguments in the most efficient way. Particularly for sympy, `add` is a lot faster than `Add` (but as a tradeoff will not perform any symbolic simplifications of the result).

**Comment by cdiener** [comment]
@KristianJensen I think the only not-so-standard ones would be `sympy.Bacic` and `sympy.core.singleton.S` if there is a `optlang.symbolics.Basic` I think we can just go ahead and try to change all the imports.

**Comment by KristianJensen** [comment]
I haven't implemented `optlang.symbolics.Basic`, but there is a `symengine.Basic` so that should be trivial. This is used with `isinstance` to check whether something is an expression I assume?`

As far as I'm aware symengine does not use singletons, so there is no equivalent to `sympy.core.singleton.S`, but is there anything in `singleton.S` that cannot be achieved otherwise? Incidentally `optlang.symbolics` will act a bit as `sympy.core.singleton.S` since e.g. `Zero` and  `One` are only instantiated once, when the module is loaded.

**Comment by KristianJensen** [comment]
I haven't implemented `optlang.symbolics.Basic`, but there is a `symengine.Basic` so that should be trivial. This is used with `isinstance` to check whether something is an expression I assume?

As far as I'm aware, symengine does not use singletons, so there is no equivalent to `sympy.core.singleton.S`, but is there anything in there that cannot be achieved otherwise? Incidentally `optlang.symbolics` will act a bit as `sympy.core.singleton.S` as e.g. `Zero` and `One` will only be instantiated once, when the module is loaded.

**Comment by cdiener** [comment]
No problem there. We only use Zero and One from S. So the changes are straightforward...

**Comment by cdiener** [comment]
Okay after some fighting I could get it to work... There are some things in symengine that are not compatible with sympy like:

- equalities between expressions (see https://github.com/symengine/symengine.py/issues/200)
- `expression.as_coefficients_dict` does not behave the same

Also I send a PR in https://github.com/biosustain/optlang/pull/134 @KristianJensen to fix some issues in optlang.

Speed-up is pretty decent. The entire suit runs about 30% faster for me. Some functions get more than that (everything that uses complex objectives more or less).

First column is sympy, second is symengine.

<details>

```bash
                                               test  time [ms] without-cache  time [ms] without-cache  fraction
1                  test_pfba_benchmark[optlang-cplex]              1870.815621               257.684712  0.137739
6            test_single_gene_deletion_moma_benchmark               504.874975               114.392641  0.226576
7          test_single_deletion_linear_moma_benchmark               265.096404                77.311789  0.291637
18                     test_loopless_benchmark_before                 4.374493                 1.806422  0.412944
31             test_copy_benchmark_large_model[cglpk]               539.910086               232.002359  0.429706
30                         test_copy_benchmark[cglpk]                20.889188                10.462938  0.500878
17                     test_legacy_loopless_benchmark                 7.815474                 4.277635  0.547329
9   test_single_gene_deletion_benchmark[optlang-cp...               184.937659               108.156195  0.584825
0                   test_pfba_benchmark[optlang-glpk]               319.278402               187.480563  0.587201
4   test_single_gene_deletion_fba_benchmark[optlan...               195.872141               123.389758  0.629951
33             test_change_objective_benchmark[cglpk]                 0.324616                 0.213107  0.656489
29          test_add_remove_reaction_benchmark[cglpk]                 0.482644                 0.320907  0.664893
25                                test_benchmark_read                15.319752                10.195421  0.665508
16  test_flux_variability_loopless_benchmark[optla...               591.784041               439.692446  0.742995
15  test_flux_variability_loopless_benchmark[optla...               589.410252               446.369807  0.757316
32                            test_deepcopy_benchmark                30.158270                23.197292  0.769185
21                           test_achr_init_benchmark               109.010914                89.959115  0.825230
22                          test_optgp_init_benchmark               108.941515                90.143068  0.827445
19                      test_loopless_benchmark_after                 1.855578                 1.610469  0.867907
12      test_flux_variability_benchmark[optlang-glpk]              1733.837673              1569.183429  0.905035
10         test_single_gene_deletion_benchmark[cglpk]                14.248036                13.054262  0.916215
2                          test_pfba_benchmark[cglpk]               391.283497               372.902529  0.953024
20               test_phenotype_phase_plane_benchmark                30.130854                28.899790  0.959143
5      test_single_gene_deletion_fba_benchmark[cglpk]                17.351443                16.684226  0.961547
26                               test_benchmark_write                32.055959                30.933192  0.964975
13     test_flux_variability_benchmark[optlang-cplex]              1619.723897              1565.524123  0.966538
28          test_subtract_metabolite_benchmark[cglpk]                 0.006109                 0.005941  0.972530
3   test_single_gene_deletion_fba_benchmark[optlan...                28.520525                27.908607  0.978545
23                         test_achr_sample_benchmark                 0.250847                 0.246255  0.981697
24                        test_optgp_sample_benchmark                 0.360591                 0.355917  0.987038
8   test_single_gene_deletion_benchmark[optlang-glpk]                21.330608                21.156021  0.991815
27               test_add_metabolite_benchmark[cglpk]                 9.633842                 9.563471  0.992695
11                test_double_gene_deletion_benchmark               771.733815               786.883310  1.019630
14             test_flux_variability_benchmark[cglpk]              1226.365009              1285.965741  1.048600
```
</details>

**Comment by pstjohn** [comment]
Is [fix/symengine](https://github.com/cdiener/cobrapy/tree/fix/symengine) ready for a PR?

**Comment by cdiener** [comment]
Still waiting for a new release of optlang since there was a small problem (fix already merged). As soon as it hits 1.2.5 I'll send the PR.

**Comment by phantomas1234** [comment]
@cdiener 1.2.5 is out

**Comment by cdiener** [comment]
Fixed in #613 .
